### PR TITLE
tf.reshape accepts Dimension objects for the shape parameter

### LINKED ISSUE #11974: tf.reshape does not accept Dimension objects for the shape parameter
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug or a feature request.
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10 x64
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**:  1.2.1
- **Python version**: 3.6.1 (Anaconda 4.4.0 64-bit)
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**: 
- **GPU model and memory**: GTX 780
- **Exact command to reproduce**: tf.reshape()

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

tf.reshape does not accept a list with mixed Integer and Dimension() as elements for the shape parameter. It should accept shapes that have Dimension as elements since tensor shapes consist of dimensions. Specifically, tf.tensor.shape returns a list of Dimensions, therefore using a similar object to specify a shape in tf.reshape should not cause an error.

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

Example code:

```
X = tf.placeholder("float", [1,784])
X = tf.reshape(X, [1, X.shape[-1]])
```

results in

```
TypeError: Failed to convert object of type <class 'list'> to Tensor. Contents: [1, Dimension(784)]. Consider casting elements to a supported type.
```

This is rectified by casting X.shape[-1] to int before passing to tf.reshape.

A similar unrelated issue is that X.shape returns (784,) while it should return (1,784). This requires reshaping to turn the placeholder back into a 2D tensor. I haven't determined if this is a bug but it occurs when the tensor is explicitly specified as a 2D tensor so it is probably worth changing.

### PR DESCRIPTION
**Pull request by facaiy**
The PR aims to fix #11974.

Because it's my first contribution, so the PR is opened early to get feedback from community.

### What changes were proposed in this pull request?

Cast tf.Dimension to int for shape argument.

### How was this patch tested?

+ [x] add a doctest.

### CODE REVIEW

**Comment by yaroslavvb** [comment]
Sorry I don't understand the motivation
compat.as_bytes(x.__str__()) instead of compat.as_bytes(x)

what does this improve?

**Comment by facaiy** [comment]
@yaroslavvb , the PR is modified as suggested by @pmccarter . But I am not sure whether it is implemented correctly, could you give more feedback? Thanks.

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Comment by rmlarsen** [comment]
@facaiy As you can see here: https://ci.tensorflow.org/job/tensorflow-pull-requests-cpu/6320/consoleFull
this breaks a large number of tests. I'll let @martinwicke give detailed comments.

**Comment by facaiy** [comment]
@rmlarsen @martinwicke  It'll be really useful for if I can check test.log of CI. Because my laptop is left on office, hence I'll check it later. Thanks very much!

**Comment by facaiy** [comment]
As `reshape` is migrated from gen_array_ops to array_ops, I guess that many unit tests will be broken because of import error.

File: tensorflow/python/ops/array_ops.py [hunk 1]
```diff
@@ -2386,6 +2387,90 @@ def where(condition, x=None, y=None, name=None):
     raise ValueError("x and y must both be non-None or both be None.")
 
 
+# pylint: disable=redefined-builtin,protected-access
```
Comment (facaiy) [inline tensorflow/python/ops/array_ops.py]: Is `protected-access` right?

**Comment by facaiy** [comment]
@tensorflow-jenkins test this please

**Comment by facaiy** [comment]
Hi, @rmlarsen . I submit a fix, and could you help retest the PR? Thanks.

File: tensorflow/python/ops/array_ops.py [hunk 2]
```diff
@@ -2386,6 +2387,88 @@ def where(condition, x=None, y=None, name=None):
...
+  Args:
+    tensor: A `Tensor`.
+    shape: A `Tensor`. Must be one of the following types: `int32`, `int64`.
+      Defines the shape of the output tensor.
+    name: A name for the operation (optional).
+
+  Returns:
+    A `Tensor`. Has the same type as `tensor`.
+  """
+  if hasattr(shape, "__len__") and len(shape) <= 8:
```
Comment (facaiy) [inline tensorflow/python/ops/array_ops.py]: Should we raise an exception when `len(shape) > 8`?

**Comment by facaiy** [comment]
By the way, since `reshape` is migrated from `_gen_array_op` to `array_op`, should we remove the duplicate document of `_gen_array_op._reshape`? If I understand correctly, the description is in the [tensorflow/core/ops/ops.pbtxt](https://github.com/tensorflow/tensorflow/blob/ca1f9846a1fa6056c28b6512e3dd1037132ef7af/tensorflow/core/ops/ops.pbtxt#L19593)

**Comment by yaroslavvb** [comment]
@facaiy The docs are inside cc file which defines reshape op.

To think of it, this approach would trigger an API review. I wonder if there's a simple solution which would only touch a couple of lines of Python code, better chance of being accepted without a lot of work on maintaining team. BTW, the best way to test things is to run all relevant tests locally first, using "bazel test"

**Comment by facaiy** [comment]
Thanks for your explanation, @yaroslavvb . 

I agree that migrating `reshape` will introduce duplicate code and increase maintain cost, so I'd like to try to fix it in `tensor_util.make_tensor_proto` as well. If success, another PR will be opened.

Running all test locally is time-consuming, hence only several unit tests are selected by me to be tested at first. By the way, could you tell me how to run doctest? Thanks.

**Pull request closed without merging**